src/main/scala/is/hail/expr/Type.scala

```diff
@@ -994,16 +1023,22 @@ final case class TDict(keyType: Type, valueType: Type) extends TContainer {
     else
       sb.append(", ")
     valueType.pretty(sb, indent, printAttrs, compact)
+    if (elementsRequired)
+      sb.append("!")
     sb.append("]")
   }
 
   def typeCheck(a: Any): Boolean = a == null || (a.isInstanceOf[Map[_, _]] &&
-    a.asInstanceOf[Map[_, _]].forall { case (k, v) => keyType.typeCheck(k) && valueType.typeCheck(v) })
+    a.asInstanceOf[Map[_, _]].forall { case (k, v) => keyType.typeCheck(k) && valueType.typeCheck(v) }) &&
+    (!elementsRequired || a.asInstanceOf[Map[_,_]].forall({case (k,v) => {assert(k != null); true}}))
```

Comment:
Put the assert inside the previous forall:

```
  a.asInstanceOf[Map[_, _]].forall { case (k, v) =>
    assert(!elementsRequired || k !=  null)
    keyType.typeCheck(k) && ...
```

Actually, on second thought, `k` isn't the element type, the pair `(k, v)`, or rather the struct `Struct { key: keyType, value: valueType }` is the element.  By definition of Map, this can never be null, so I think there's nothing to check here and we should have `TDict.elementsRequired = true`.

Reply:
Separately we might want to require the keys (or values) to be non-missing, like `Dict[K!, V]` or `Dict[K!, V!]` but I say we don't support this for now.  (Or do support it.  But just make sure you distinguish the key from the value from the element.)